pkg/registry/generic/matcher.go

```diff
@@ -48,78 +43,4 @@ func MergeFieldsSets(source fields.Set, fragment fields.Set) fields.Set {
 	return source
 }
 
-// SelectionPredicate implements a generic predicate that can be passed to
-// GenericRegistry's List or Watch methods. Implements the Matcher interface.
-type SelectionPredicate struct {
-	Label       labels.Selector
-	Field       fields.Selector
-	GetAttrs    AttrFunc
-	IndexFields []string
-}
-
-// Matches returns true if the given object's labels and fields (as
-// returned by s.GetAttrs) match s.Label and s.Field. An error is
-// returned if s.GetAttrs fails.
-func (s *SelectionPredicate) Matches(obj runtime.Object) (bool, error) {
-	if s.Label.Empty() && s.Field.Empty() {
-		return true, nil
-	}
-	labels, fields, err := s.GetAttrs(obj)
-	if err != nil {
-		return false, err
-	}
-	matched := s.Label.Matches(labels)
-	if s.Field != nil {
-		matched = (matched && s.Field.Matches(fields))
-	}
-	return matched, nil
-}
-
-// MatchesSingle will return (name, true) if and only if s.Field matches on the object's
-// name.
-func (s *SelectionPredicate) MatchesSingle() (string, bool) {
-	// TODO: should be namespace.name
-	if name, ok := s.Field.RequiresExactMatch("metadata.name"); ok {
-		return name, true
-	}
-	return "", false
-}
-
-// For any index defined by IndexFields, if a matcher can match only (a subset)
-// of objects that return <value> for a given index, a pair (<index name>, <value>)
-// wil be returned.
-// TODO: Consider supporting also labels.
-func (s *SelectionPredicate) MatcherIndex() []storage.MatchValue {
-	var result []storage.MatchValue
-	for _, field := range s.IndexFields {
-		if value, ok := s.Field.RequiresExactMatch(field); ok {
-			result = append(result, storage.MatchValue{IndexName: field, Value: value})
-		}
-	}
-	return result
-}
-
-// Matcher can return true if an object matches the Matcher's selection
-// criteria. If it is known that the matcher will match only a single object
-// then MatchesSingle should return the key of that object and true. This is an
-// optimization only--Matches() should continue to work.
-type Matcher interface {
-	// Matches should return true if obj matches this matcher's requirements.
-	Matches(obj runtime.Object) (matchesThisObject bool, err error)
-
-	// If this matcher matches a single object, return the key for that
-	// object and true here. This will greatly increase efficiency. You
-	// must still implement Matches(). Note that key does NOT need to
-	// include the object's namespace.
-	MatchesSingle() (key string, matchesSingleObject bool)
-
-	// For any known index, if a matcher can match only (a subset) of objects
-	// that return <value> for a given index, a pair (<index name>, <value>)
-	// will be returned.
-	MatcherIndex() []storage.MatchValue
-}
-
-var (
-	// Assert implementations match the interface.
-	_ = Matcher(&SelectionPredicate{})
-)
+type SelectionPredicate storage.SelectionPredicate
```

Comment:
This alias could be removed.
I was keeping it here for review's sake -- otherwise there will be a lot more changes.
Will change it shortly